kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -122,12 +122,18 @@ public suspend inline fun <T> Mutex.withLock(owner: Any? = null, action: () -> T
         callsInPlace(action, InvocationKind.EXACTLY_ONCE)
     }
 
+    // Cannot use 'finally' in this function because of KT-58685
+    // See kotlinx.coroutines.sync.MutexTest.testWithLockJsMiscompilation
+
     lock(owner)
-    try {
-        return action()
-    } finally {
+    val result = try {
+        action()
+    } catch (e: Throwable) {
         unlock(owner)
+        throw e
     }
```

Comment:
To recreate the behavior of `finally`, if `unlock` inside `catch` throws an exception, this exception must add `e` as a suppressed one. Could you also write the corresponding test?

Reply:
Hi, sorry for the time it took me to get back to this.

I'm not sure exactly what you mean here, something like this?
```kotlin
try {
    action()
} catch (e: Throwable) {
    try {
        unlock(owner)
    } catch (suppressed: Throwable) {
        e.addSuppressed(suppressed)
    }
    throw e
}
```

Reply:
Oh, sorry, I was mistaken. There's no such behavior currently, and on second thought, I don't think there should be.